[quant] Add custom freeze pass for quantization attributes

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43595 [quant] Expose quantization freeze pass in convert API
* **#43594 [quant] Add custom freeze pass for quantization attributes**

Summary:

This PR decouples quantization finalize pass from the geenric `freeze_module` pass
It only focuses on freezing the attributes related to quantization params so that they can later
be folded into the graph, and used by subsequent JIT pass like prepack_folding

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23333263](https://our.internmc.facebook.com/intern/diff/D23333263)

supriyar committed:
[quant] Add custom freeze pass for quantization attributes

Summary:

This PR decouples quantization finalize pass from the geenric `freeze_module` pass
It only focuses on freezing the attributes related to quantization params so that they can later
be folded into the graph, and used by subsequent JIT pass like prepack_folding

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/finalize.cpp:
```diff
@@ -73,12 +74,31 @@ void FoldQuantizedPrepackingOps(Module& module) {
-Module Finalize(Module& module, QuantType quant_type) {
+Module FreezeQuantizedAttrs(Module& module) {
+  auto filter_fn = [](const Node* n) -> bool {
+    return (
```
optional nit: maybe put the symbols in a list and use find to check this.

jerryzh168 commented on torch/csrc/jit/passes/quantization/freeze_quant_ops.cpp:
```diff
@@ -0,0 +1,201 @@
+    auto graph = func_->graph();
+    // Inline the graph to make sure we don't have to traverse different
+    // subgraphs when trying to get the weight attributes.
+    Inline(*graph);
```
can we inline?

jerryzh168 commented on torch/csrc/jit/passes/quantization/freeze_quant_ops.cpp:
```diff
@@ -0,0 +1,201 @@
+      if (quant_inp_node->kind() == prim::GetAttr) {
+        convertAttrToConst(module, quant_inp_node, graph);
+      } else if (
+          quant_inp_node->kind() == Symbol::aten("quantize_per_tensor") ||
+          quant_inp_node->kind() == Symbol::aten("quantize_per_channel")) {
```
maybe don't need this check?

jerryzh168 commented on torch/csrc/jit/passes/quantization/freeze_quant_ops.cpp:
```diff
@@ -0,0 +1,201 @@
+    auto attrModule = module;
+    std::deque<std::string> module_names;
+    // Get the module corresponding to the attr
+    if (!updateAttrModule(
```
I think we have helper function for this? https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit16getInvokedModuleERNS0_6ModuleEPNS0_4NodeEPNS0_5ValueE

supriyar committed:
Update on "[quant] Add custom freeze pass for quantization attributes"


Summary:

This PR decouples quantization finalize pass from the geenric `freeze_module` pass
It only focuses on freezing the attributes related to quantization params so that they can later
be folded into the graph, and used by subsequent JIT pass like prepack_folding

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23333263](https://our.internmc.facebook.com/intern/diff/D23333263)

[ghstack-poisoned]

supriyar commented:
This is not needed anymore. Closing.
